File: Array/buySellStock.py
#given an array prices where prices[i] is the price of a given stock on the ith day.
#maximize your profit by choosing a single day to buy one stock 
#and choosing a different day in the future to sell that stock.
#Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

prices = [20, 9, 11, 40, 71]
min_price = prices[0]
min_price_index = 0

for index,price in enumerate(prices):
    if price < min_price:
        min_price = price
        min_price_index = index

print("The buying price is ", min_price)

#check if our smallest price is on the last index of the list
if min_price_index < len(prices) - 1:
      max_price =max(prices[min_price_index + 1:])
      if max_price > min_price:
          max_profit = max_price - min_price
          print(f"Buy the stock at {min_price} on day {min_price_index + 1 } and sell it at {max_price} with {max_profit} as profit")
      else:
          print(0)
else:
    print(0)
   

File: Array/buySellStock.js
// given an array of prices input
// maximazie to profit on selecting a specific day to buy and sell 
// return maximum profit
// you must sell before you buy

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let min_price = prices[0];
    let maxProfit = 0;
    for (let i = 0; i < prices.length; i++) {
        // Update the minimum price if we find a lower price
        min_price = Math.min(min_price, prices[i]);
        // Update the maximum profit if selling at the current price gives a higher profit
        maxProfit = Math.max(maxProfit, prices[i] - min_price);
    }
    return maxProfit;
};
const prices = [7,6,5,4,2,3];
const maxiProfit = maxProfit(prices)

console.log(maxiProfit);